core/src/mender-client.c

```diff
@@ -83,23 +83,26 @@ mender_client_state_t mender_client_state = MENDER_CLIENT_STATE_INITIALIZATION;
     mender_update_state_t failure;
+    mender_update_state_t state_loop;
 };
 
 /**
  * @brief Mender Update (module) state transitions
  */
+/* clang-format off */
 static const struct mender_update_state_transition_s update_state_transitions[N_MENDER_UPDATE_STATES] = {
-    /* MENDER_UPDATE_STATE_DOWNLOAD               */ { MENDER_UPDATE_STATE_INSTALL, MENDER_UPDATE_STATE_CLEANUP },
-    /* MENDER_UPDATE_STATE_INSTALL                */ { MENDER_UPDATE_STATE_REBOOT, MENDER_UPDATE_STATE_FAILURE },
-    /* MENDER_UPDATE_STATE_REBOOT                 */ { MENDER_UPDATE_STATE_VERIFY_REBOOT, MENDER_UPDATE_STATE_ROLLBACK },
-    /* MENDER_UPDATE_STATE_VERIFY_REBOOT          */ { MENDER_UPDATE_STATE_COMMIT, MENDER_UPDATE_STATE_ROLLBACK },
-    /* MENDER_UPDATE_STATE_COMMIT                 */ { MENDER_UPDATE_STATE_CLEANUP, MENDER_UPDATE_STATE_ROLLBACK },
-    /* MENDER_UPDATE_STATE_CLEANUP                */ { MENDER_UPDATE_STATE_END, MENDER_UPDATE_STATE_END },
-    /* MENDER_UPDATE_STATE_ROLLBACK               */ { MENDER_UPDATE_STATE_ROLLBACK_REBOOT, MENDER_UPDATE_STATE_FAILURE },
-    /* MENDER_UPDATE_STATE_ROLLBACK_REBOOT        */ { MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT, MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT },
-    /* MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT */ { MENDER_UPDATE_STATE_FAILURE, MENDER_UPDATE_STATE_ROLLBACK_REBOOT },
-    /* MENDER_UPDATE_STATE_FAILURE                */ { MENDER_UPDATE_STATE_CLEANUP, MENDER_UPDATE_STATE_CLEANUP },
+    /* MENDER_UPDATE_STATE_DOWNLOAD               */ { MENDER_UPDATE_STATE_INSTALL, MENDER_UPDATE_STATE_CLEANUP, MENDER_UPDATE_STATE_FAILURE },
+    /* MENDER_UPDATE_STATE_INSTALL                */ { MENDER_UPDATE_STATE_REBOOT, MENDER_UPDATE_STATE_FAILURE, MENDER_UPDATE_STATE_FAILURE },
+    /* MENDER_UPDATE_STATE_REBOOT                 */ { MENDER_UPDATE_STATE_VERIFY_REBOOT, MENDER_UPDATE_STATE_ROLLBACK, MENDER_UPDATE_STATE_FAILURE },
+    /* MENDER_UPDATE_STATE_VERIFY_REBOOT          */ { MENDER_UPDATE_STATE_COMMIT, MENDER_UPDATE_STATE_ROLLBACK, MENDER_UPDATE_STATE_FAILURE },
+    /* MENDER_UPDATE_STATE_COMMIT                 */ { MENDER_UPDATE_STATE_CLEANUP, MENDER_UPDATE_STATE_ROLLBACK, MENDER_UPDATE_STATE_FAILURE },
+    /* MENDER_UPDATE_STATE_CLEANUP                */ { MENDER_UPDATE_STATE_END, MENDER_UPDATE_STATE_END, MENDER_UPDATE_STATE_END },
+    /* MENDER_UPDATE_STATE_ROLLBACK               */ { MENDER_UPDATE_STATE_ROLLBACK_REBOOT, MENDER_UPDATE_STATE_FAILURE, MENDER_UPDATE_STATE_FAILURE },
+    /* MENDER_UPDATE_STATE_ROLLBACK_REBOOT        */ { MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT, MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT, MENDER_UPDATE_STATE_FAILURE },
+    /* MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT */ { MENDER_UPDATE_STATE_FAILURE, MENDER_UPDATE_STATE_ROLLBACK_REBOOT, MENDER_UPDATE_STATE_FAILURE },
+    /* MENDER_UPDATE_STATE_FAILURE                */ { MENDER_UPDATE_STATE_CLEANUP, MENDER_UPDATE_STATE_CLEANUP, MENDER_UPDATE_STATE_CLEANUP },
```

Comment:
I'll let @vpodzime comment on this part - maybe it is not worth to have a third parameter in the transition struct if it's only two states that have something different than `FAILURE` (iow maybe is better to have to if/else if for these)

Reply:
Yes. And there is no way the two special states will loop, right?

Reply:
They won't loop - all states except `failure` and `cleanup` goes to `failure`; `failure` goes to `cleanup`, and `cleanup` always goes to `end`. But you don't think it should be a part of the transition struct? Isn't it cleaner/clearer if we have the possible transitions in one place?

Reply:
I'm thinking a loop should always be quit by going to the `failure` state with the counter reset. That way we hopefully report an error to the server, do a cleanup and return back to normal operation.

Reply:
Right - so just `if loop-detected: update_state = failure` then I would agree that it's not necessary to have it in the struct - if that's what you mean

Reply:
> there is no point in going to failure if we're in the cleanup state

Right, but there can never be a loop when we entered the `cleanup` state, right? And the same applies to `failure`. That's what in meant [above](https://github.com/mendersoftware/mender-mcu/pull/96#discussion_r1808335109)

Reply:
Maybe I am misunderstanding you, but I do believe it's possible that there somehow is a back-and-forth between some states which can result in cleanup being entered right at the state-loop threshold, resulting in a loop being detected while we're entering cleanup. Same with failure - since both states call NEXT_STATE, which attempts to store the deployment data and increases the state loop counter. So we would probably need to store which state we're in before calling transition.success or transition.failure, then have logic to not enter failure in failure and cleanup. It isn't hard to implement, but I feel like that creates more clutter in `NEXT_STATE` rather than simply having a third column in the transition struct :sweat_smile:  - or maybe I am just overlooking a much simpler solution

Reply:
> I do believe it's possible that there somehow is a back-and-forth between some states which can result in cleanup being entered right at the state-loop threshold
Then we have a bad threshold because it's not a loop, right? But if we just go to `failure` every time we detect a loop and reset the counter, the worst that can happen is that we:
1. detect a loop while entering `cleanup`
2. go to `failure` and reset the counter
3. go to `cleanup` with counter at 1

(or the same for detecting loop when entering `failure`).

Is that not going to work in some case? `failure` an `cleanup` cannot create a loop on their own, right?

Reply:
That will indeed work - didn't even think about just resetting the counter, thanks :)